### ISSUE TITLE
password recovery feature with datamaper issue

### ISSUE THREAD

**Issue by tigretoncio** [makersacademy/slack-overflow#199]
Hello,

I have followed the following walkthrough [password recovery](https://github.com/makersacademy/course/blob/master/bookmark_manager/walkthroughs/26.md)
but I can't make it work.  I have tried to debug the relevant issues, and got to a situation in which I get the following error:

`undefined method `+' for nil:NilClass`

`BACKTRACE
(expand)
JUMP TO: GET POST COOKIES ENV
/home/sergio/Ronin/Week4/bookmark_manager/app/models/user.rb in find_by_valid_token
    if (user && user.password_token_time + (60 * 60) > Time.now)
/home/sergio/Ronin/Week4/bookmark_manager/app/controllers/users.rb in block in <class:BookmarkManager>
     @user = User.find_by_valid_token(params[:token])
`

The find_by_valid_token is as follows:

````
  def self.find_by_valid_token(token)
    user = first(password_token: token)
    if (user && user.password_token_time + (60 * 60) > Time.now)
      user
    end
  end
````

If I correctly understand, the error is because user is nil, and therefore nil + something is not defined.  However I am confused because the condition (user && ....) should evaluate user as nil and stop the evaluation at that point and return false.  It is not the case, so perhaps user.password_token_time is nul but not user...

Is there any way of showing the value of user at this moment in the program using this modular programming?

Have anyone got a version of this password recovery working? I would like to understand what I am doing wrong, or where the walkthrough is not accurate.

Any help or hint or breadcrumb appreciated...

For reference, some of my controllers and views:

controllers/users.rb
````
class BookmarkManager < Sinatra::Base
  get '/users/new' do
    @user = User.new
    erb :'users/new'
  end

  post '/users' do
    @user = User.new(email: params[:email],
     password: params[:password],
     password_confirmation: params[:password_confirmation])
    if @user.save
      session[:user_id] = @user.id
      redirect to('/links')
    else
      #flash.now[:users] = @user.errors.full_messages
      erb :'users/new'
    end
  end

  get '/users/recover' do
    "Please enter your email address"
    erb :'users/recover'
  end

  post '/users/recover' do
    @user = User.first(email: params[:email])
    if @user
      @user.generate_token
    end
    erb :'users/acknowledgment'
  end

  get '/users/reset_password' do
    @user = User.find_by_valid_token(params[:token])
    if(@user)
      erb :'users/reset_password'
    else
      "Your token is invalid"
    end
  end

   patch '/users' do
     @user = User.find_by_valid_token(params[:token])
     if (@user)
       @user.update(password: params[:password], password_confirmation: params[:password_confirmation])
       redirect '/sessions/new'
     else
       flash.now[:errors] = "Your token is invalid"
       redirect 'users/reset_password'
     end
   end
end
````

views/users/recover.erb
````
<form action='/users/recover' method='post'>
  <label for='email'> Please enter your email address
    Email: <input name='email' type='email' required>
  </label>
  <input type='submit' value='Submit'>
</form>
````

views/users/reset_password.erb
````
<h1>Please enter your new password</h1>

<form action='/users' method='post'>
 <input type="hidden" name="_method" value="patch">
 <label for='password'>
   Password: <input name='password' type='password'>
 </label>
 <label for='password_confirmation'>
   Password Confirmation: <input name='password_confirmation' type='password'>
 </label>
 <input type='submit' value='Submit'>
</form>
````

views/users/acknowledgment.erb
(the variables shown are for debugging purposes, I can see token is correctly generated and 
can go onto the page to put a new password, then when I click submit I get the nil error)
````
Thanks, Please check your inbox for the link.
Generated token:
<%= @user.password_token %>  
<br>
<%= @user.password_token_time %>
````

And the model
models/user.rb
````
require 'securerandom'

class User
  include DataMapper::Resource

  property :id,               Serial
  property :count,            Integer
  property :email, String, required: true, format: :email_address, unique: true
  property :password_digest,  Text
  property :password_token, Text
  property :password_token_time, Time

  attr_reader :password
  attr_accessor :password_confirmation

  validates_confirmation_of :password


  def password=(password)
  	@password = password
    self.password_digest = BCrypt::Password.create(password)
  end

  def self.authenticate(email, password)
    user = first(email: email)
    if user && BCrypt::Password.new(user.password_digest) == password
      user
    else
      nil
    end
  end

  def generate_token
    self.password_token = SecureRandom.hex
    self.password_token_time = Time.now
    self.save
  end

  def self.find_by_valid_token(token)
    user = first(password_token: token)
    if (user && user.password_token_time + (60 * 60) > Time.now)
      user
    end
  end
end
````

**Comment by tigretoncio** [comment]
zero answers in 14 days... is it because my question is wrongly worded?

**Comment by tigretoncio** [comment]
TL;DR...
